Reimplement Coraza-SPOA

fionera opened the pull request:
Another go at updating coraza-spoa to a the up to date coraza version and a faster spop library. It also uses upstream e2e tests and 
---

Q: Why rewrite instead of incremental changes?
A: I didn't had time nor motivation to figure out why some things are how they are. It was just easier to do it like this.

---

Other PRs that are similar:
- #90
- #89

jptosso commented:
Lgtm

jcchavezs commented on .github/workflows/container-image.yaml:
```diff
@@ -30,14 +21,6 @@ jobs:
       - name: Check out code
         uses: actions/checkout@v3
 
-      - name: Run e2e tests against the example
```
I'd comment this out rather than removing them. We indeed need these e2e test to make sure the whole integration works.

jcchavezs replied:
Closing the comment as I see you turned e2e tests into go code :).

jcchavezs commented on .github/workflows/test.yaml:
```diff
@@ -0,0 +1,27 @@
+        with:
+          go-version: '1.21'
+
+      - name: setup environment
```
I'd really avoid going in this direction, this assumes a haproxy is running in local which is going to be a big friction to those willing to contribute. Personally I would never install haproxy locally to debug a problem so I suggest we spin up a haproxy instance using docker with `testcontainers` api (see https://pkg.go.dev/github.com/testcontainers/testcontainers-go) so the test is selfcontained and does not assume I have something installed.

fionera replied:
I explicitly did this without using containers (atleast inside the CI) as its faster to run. Locally and in production I would always recommend running without docker as the performance difference is huge.

fionera replied:
(Also I dont have docker on my machine since its just very slow on macos)

jcchavezs replied:
I really prefer to have something that is portable. It is OK if we want CI to run haproxy locally but we also need to provide a way to run it with docker. Again here we prioritize user experience if I have to install locally haproxy to contribute I'd rather not contribute hence we use containers.

jcchavezs commented on internal/e2e_test.go:
```diff
@@ -0,0 +1,132 @@
+	"github.com/dropmorepackets/haproxy-go/pkg/testutil"
+)
+
+const directives = `
```
Please use `go:embed` for this, also in general we would like to keep a consistent http e2e suite by suing https://github.com/corazawaf/coraza/tree/main/http/e2e which is being used in coraza http handler, caddy and coraza-proxy-wasm. Running it is as easy as spin up an instance and run a go command, see https://github.com/corazawaf/coraza-caddy/blob/main/magefile.go#L74

fionera replied:
Why should I use embed for a single string? Inside the coraza tests its done the same way.

fionera replied:
But yeah I can migrate the e2e tests to run the command but I am not a fan of running commands and basically ignoring the go test framework. If the Upstream would properly expose the tests as a list we could run these in parallel and reduce testing times while also reducing the amount of external dependencies

fionera replied:
I opened an issue on coraza itself to expose this and make it easier for other repos to use the e2e test suite.
https://github.com/corazawaf/coraza/issues/1006

fionera replied:
Waiting for a coraza release to fix this up

jcchavezs replied:
Maybe it is better you use master for now.

jcchavezs commented on internal/application.go:
```diff
@@ -0,0 +1,326 @@
+		case "method":
+			req.Method = string(k.ValueBytes())
+		case "path":
+			// make a copy of the pointer and add a defer in case there is another entry
```
Is this possible? can you point out some references on this behaviour?

jcchavezs commented on internal/application.go:
```diff
@@ -0,0 +1,326 @@
+		return fmt.Errorf("request id is empty")
+	}
+
+	tx := a.waf.NewTransactionWithID(req.ID)
```
I am usually reluctant about this, is req.ID something unique per instance e.g. in a distributed environment all hops in a request will reuse the X-Request-Id, not sure if this is the same thing or is it a unique value generated on every request to haproxy.

fionera replied:
This is the same behaviour as current [main](https://github.com/corazawaf/coraza-spoa/blob/main/internal/spoa.go#L279). The request ID is set via [unique-id-format](https://docs.haproxy.org/dev/configuration.html#4-unique-id-format) in HAProxy [here (e2e)](https://github.com/corazawaf/coraza-spoa/blob/rewrite/internal/e2e_test.go#L79) and [here (examples)](https://github.com/corazawaf/coraza-spoa/blob/rewrite/examples/haproxy.cfg#L19).

jcchavezs replied:
Do you know if haproxy propagates it?

fionera replied:
Just to be on the same page, the Header will be injected into the request and send to the backend. I dont know what you mean with distributed environments

fionera replied:
Based on an idea by @mxey I moved the id generation to the request handler, set it as spop transaction value and pass it back inside the response handler. This removes the requirement of `unique-id-format`

fionera commented:
Sorry that I didn't update on this again but I am currently working on getting a talk done. Will try to get the other points done by the end of next week

fionera commented:
I think this can be reviewed and tested now. There is still a fd leak when reloading and changing the logger, also when there is a reload with logger change and then an a reload that reverts this, the old file can be raced. There should probably a small wrapper around the logger to provide prevent multiple writers to the same file, but I guess this can wait.

fionera commented:
Sure will do that

fionera commented:
Updated :)

fionera commented:
And now without missing to remove an old file 🗡️

csuka commented:
Yeah, can we ensure this gets approved/merged?

superstes commented:
FYI: I tested the version of this PR and it fixed https://github.com/corazawaf/coraza-spoa/issues/109 for me.

superstes commented:
FYI: This version is working fine for me. Any reason for the delay?

fzipi commented:
@jcchavezs Can you engage here?

superstes commented on internal/agent.go:
```diff
@@ -0,0 +1,96 @@
+	if !k.NameEquals("app") {
+		// Without knowing the app, we cannot continue. We could fall back to a default application,
+		// but all following code would have to support that as we now already read one of the kv entries.
+		a.Logger.Panic().Str("expected", "app").Str("got", appName).Msg("unexpected kv entry")
```
I often see the error `unexpected kv entry expected=app got=sample_app` when configuring the app-name `sample_app`. Could this be an issue? I'm not sure why this happens

fionera replied:
Ah yeah this should not print the field value but the field name

fionera replied:
Can you check now what the correct error is? The fields in the haproxy/coraza.cfg are expected to be in that specific order

superstes replied:
Alright. With the new version I see: `unexpected kv entry expected=app got=path`
Have the exact `coraza.cfg` from your examples: (and working most of the time)
```
spoe-message coraza-req
    # Arguments are required to be in this order
    args app=str(sample_app) src-ip=src src-port=src_port dst-ip=dst dst-port=dst_port method=method path=path query=query version=req.ver headers=req.hdrs body=req.body
    event on-frontend-http-request
```

fionera replied:
Which HAProxy version? I would like to reproduce this.

superstes replied:
Will E-Mail you the complete config.

fionera replied:
Found the issue! There was a race inside the decoder. Fixed now :)

superstes replied:
Nice :+1: 
Changes are up in my env. Will report back tomorrow

superstes replied:
Have not yet seen that error again.
The only thing that is visible are `reading frame length: read tcp 127.0.0.1:9000->127.0.0.1:xxxxx: read: connection reset by peer` logs.
~1.000 logs in the last 24h

Not sure this is normal. If they are expected - we might want to disable these logs or configure them to be debug-logs.

fionera replied:
These are expected and happen when HAProxy closes a connection instead of reusing them. These will stay for now as I have to investigate the behaviour of haproxy regarding this, but they aren't anything bad.

fionera commented on internal/application.go:
```diff
@@ -0,0 +1,337 @@
+	AppConfig
+}
+
+type applicationRequest struct {
```
We generate this id some lines further down and assign it as variable to share it with haproxy.

superstes replied:
Thank you for the clarification. 👍

Maybe we should add a line to `haproxy.cfg` that showcases how to add this ID to the haproxy-logs.
This is currently shown in the docs: https://coraza.io/connectors/coraza-spoa/#haproxy-configuration (with log-format)

fionera replied:
Sure I will add a comment to the config

superstes replied:
Thank you 🙏

superstes commented on internal/application.go:
```diff
@@ -0,0 +1,361 @@
+	default:
+		l = a.Logger.Error()
+	}
+	l.Msg(mr.ErrorLog())
```
There might be a problem with the JSON log-format.
The spoa logs are in JSON, but the ones from coraza are not. (`SecAuditLogFormat JSON` is ignored)
It works with basic coraza 3.2.1
Some more info: https://github.com/corazawaf/coraza/issues/1150

fionera commented:
@sts @jcchavezs Can you do a review? I would love to have this merged to start working on fixing/cleaning up the other issues

nicolus commented:
I'm in no way competent to judge any of this, but as someone who's wanting to switch from modsecurity to Coraza on my HAProxy setup I'm really glad to see someone making the effort to work on this.

So yeah if you guys could take some time to review this PR 😅

mmilitzer commented:
Since we're having stability issues with the current main branch (See: https://github.com/corazawaf/coraza-spoa/issues/97) I've been following this rewrite closely hoping it would get merged and fix our issues. 

But since merging is stalled since a while I got curious and went ahead with deploying the rewrite branch anyway and putting some of our production traffic against it for a test. But unfortunately, this didn't go well - I couldn't make any conclusions on the stability vs. main because the rewrite branch has a massive memory leak issue and so I couldn't keep it running it for long. So after less than two hours the coraza-spoa daemon's memory use grew above 4 GB and additionally it also used a lot more CPU (about twice as much or more) than the main branch version. And so I ended the experiment again.

My setup uses the CRS v4.7.0 plus a few custom exclusion rules to deal with false positives. Response checking is disabled. Config options and rules were the same when doing this comparison, so the only thing that differs is that one time I ran the coraza-spoa daemon binary built from the rewrite branch and the other time that from the main branch.

First, here's a graph of how the overall system's memory consumption developed after enabling the rewrite branch coraza-spoa daemon. There's an obvious memory leak which forced me to stop the experiment and downgrade to the main branch version again after ~2 hrs. And, after switching back to main, the memory use instantly became stable again:
![System_Memory_Use](https://github.com/user-attachments/assets/3e19ed62-fc3a-4f5e-aad1-a8ae2e63a054)

And also for CPU usage there's a clear difference, so see the CPU consumption graph below that depicts the CPU usage of the coraza-spoa daemon process. While running the rewrite branch version the CPU consumption is consistently much higher and after downgrading to the main branch version it drops again to a much lower level:
![coraza-spoa-CPU_Usage](https://github.com/user-attachments/assets/b5b7cf2f-d2a0-411f-9033-3813090d7768)

What I further noticed is that I got very frequently messages like the following below in our /var/log/messages when running the rewrite branch version:

coraza-spoa[3659835]: 2024/09/28 16:13:01 write tcp 127.0.0.1:9000->127.0.0.1:50932: write: broken pipe
coraza-spoa[3659835]: 2024/09/28 16:13:01 write tcp 127.0.0.1:9000->127.0.0.1:50932: write: broken pipe
coraza-spoa[3659835]: 2024/09/28 16:13:01 write tcp 127.0.0.1:9000->127.0.0.1:50932: write: broken pipe
coraza-spoa[3659835]: 2024/09/28 16:13:01 write tcp 127.0.0.1:9000->127.0.0.1:50932: write: broken pipe
coraza-spoa[3659835]: 2024/09/28 16:13:01 write tcp 127.0.0.1:9000->127.0.0.1:50932: write: broken pipe
coraza-spoa[3659835]: 2024/09/28 16:13:01 write tcp 127.0.0.1:9000->127.0.0.1:50932: use of closed network connection
coraza-spoa[3659835]: 2024/09/28 16:13:01 write tcp 127.0.0.1:9000->127.0.0.1:50932: use of closed network connection
coraza-spoa[3659835]: 2024/09/28 16:13:01 write tcp 127.0.0.1:9000->127.0.0.1:50932: use of closed network connection
coraza-spoa[3659835]: 2024/09/28 16:13:02 write tcp 127.0.0.1:9000->127.0.0.1:50932: use of closed network connection
coraza-spoa[3659835]: 2024/09/28 16:13:02 write tcp 127.0.0.1:9000->127.0.0.1:50932: use of closed network connection

With the main branch coraza-spoa daemon we're also sometimes seeing such "broken pipe" log messages but they are very rare, so they happen just every couple of days. But while running the rewrite branch version this happened all the time, like there were some such messages logged almost every minute. So I'm wondering if there are also functional issues then (beyond just the resource consumption problems). Or if maybe these "broken pipe" errors also cause the memory leak (perhaps under these error conditions memory isn't properly released?).

But so unfortunately, I have no clue what the root cause of these issues might be. Perhaps, it's not even the code in the rewrite branch itself, so the rewrite branch also brings in a lot of newer dependencies - hence the memory leak could also be caused by a dependency (though less likely). So I can only share these observations from our first real-world test now and unfortunately it seems the rewrite branch isn't quite ready for general use yet. But hopefully these remaining issues can still be solved now...

csuka commented:
> Or if maybe these "broken pipe" errors also cause the memory leak.

Perhaps this could help you out to find the root cause of the issue:

"A "broken pipe" error typically indicates that a process was attempting to write data to another process or socket, but the receiving process was closed or became unavailable before the data transfer was complete. This can occur for various reasons, including:

Network Issues: If the communication was happening over a network, a drop in network connectivity or a closed connection can result in a broken pipe.
Process Termination: If a receiving process crashes, exits, or is terminated unexpectedly, any ongoing write operation to that process will fail with a broken pipe error.
Resource Limits: Resource limitations like memory or file descriptors reaching their maximum capacity can cause one process to close, resulting in a broken pipe.
Timeouts: The receiving process may have a timeout set and, upon expiration, could close the connection while data is still being sent."

fionera commented:
Did you check with before the fixes I pushed a couple days ago? I found the resource leak which was because of a wrong usage of defer statements. The brokenn pipe issues are happening when haproxy decides to close the connection because of either too long evaluation times or too many requests. I am currently investivating why exactly haproxy does that and will keep you updated.

mmilitzer commented:
> Did you check with before the fixes I pushed a couple days ago?

Yes, I used the latest version available (commit [6ea3e43](https://github.com/corazawaf/coraza-spoa/commit/6ea3e439c6d9beebc9e8361f84a34cc7968bfdf1))

fionera commented:
I cannot reproduce this behaviour on my side. Can you run the newest commit with the memory profile flag?

fionera commented:
> > Did you check with before the fixes I pushed a couple days ago?
> 
> Yes, I used the latest version available (commit [6ea3e43](https://github.com/corazawaf/coraza-spoa/commit/6ea3e439c6d9beebc9e8361f84a34cc7968bfdf1))

Can you try with the newest master? I fixed the disconnect handling and added a memprofile flag. If the memory leak is still happening I would love to get a pprof to debug it further. Thank you :)

mmilitzer commented:
> Can you try with the newest master? I fixed the disconnect handling and added a memprofile flag. If the memory leak is still happening I would love to get a pprof to debug it further. Thank you :)

I've tried the latest rewrite branch code and I have partially good news: The memory leak is gone, so I couldn't reproduce this anymore. Also the CPU usage was much better and on around the same level than the version from current main branch. So it looks like the memory leak previously was also eating CPU cycles and that's why CPU usage was higher. So the resource usage issues seem to be solved now, not worse than main branch anymore.

However, I'm still seeing "broken pipe" etc. output in /var/log/messages frequently which doesn't happen when running the code from the main branch:

```log
2024-09-30T17:26:00.221389+00:00 web05 coraza-spoa[257058]: 2024/09/30 17:26:00 disconnect frame with code 3: frame is too 
2024-09-30T17:26:00.221466+00:00 web05 coraza-spoa[257058]: 2024/09/30 17:26:00 disconnect frame with code 3: frame is too big
2024-09-30T17:27:46.019387+00:00 web05 coraza-spoa[257058]: 2024/09/30 17:27:46 write tcp 127.0.0.1:9000->127.0.0.1:46636: write: broken pipe
2024-09-30T17:27:46.019485+00:00 web05 coraza-spoa[257058]: 2024/09/30 17:27:46 disconnect frame with code 3: frame is too big
2024-09-30T17:27:46.019564+00:00 web05 coraza-spoa[257058]: 2024/09/30 17:27:46 write tcp 127.0.0.1:9000->127.0.0.1:46636: use of closed network connection
2024-09-30T17:29:05.046278+00:00 web05 coraza-spoa[257058]: 2024/09/30 17:29:05 disconnect frame with code 3: frame is too big
2024-09-30T17:29:05.046511+00:00 web05 coraza-spoa[257058]: 2024/09/30 17:29:05 disconnect frame with code 3: frame is too big
2024-09-30T17:29:37.683271+00:00 web05 coraza-spoa[257058]: 2024/09/30 17:29:37 disconnect frame with code 3: frame is too big
2024-09-30T17:29:37.683348+00:00 web05 coraza-spoa[257058]: 2024/09/30 17:29:37 write tcp 127.0.0.1:9000->127.0.0.1:59350: write: broken pipe
```

So this is just a short snippet from the log but it's those same three errors ("frame is too big", "use of closed network connection" and "broken pipe") that repeat again and again. And "frame is too big" seems new, so I think I didn't see this before when I tested the earlier version that still had the memory leak.

NOTE: I tested this all still with go 1.21 because I don't have a newer version on AlmaLinux 9 and was too lazy to build the newer one from source. So I've reverted your latest commit d0ba9f8ba405a8aada589de7fb2e4d6472a3843d and also changed the go version in go.mod back to 1.21 to make it work. And I don't know if that's a problem, so if there's anything why go >= 1.23.1 is really required...

superstes commented on go.mod:
```diff
@@ -1,27 +1,30 @@
 module github.com/corazawaf/coraza-spoa
 
-go 1.19
+go 1.23.1
```
With go 1.23.1 the application fails to start for me.
With 1.22.3 (as used by the waf-core https://github.com/corazawaf/coraza/blob/main/go.mod#L3) it works

Error:
`Oct 06 16:43:57 lb01 coraza[79533]: {"level":"fatal","error":"initializing application '\\x00': invalid WAF config from string: failed to readfile: read /etc/coraza-spoa/coraza.conf: invalid argument","time":"2024-10-06T16:43:57+02:00","message":"Failed creating applications"}`

dknight-bg replied:
same error with go 1.23.2 :
9:45PM FTL Failed creating applications error="initializing application '\\x00': invalid WAF config from string: failed to readfile: read /opt/coraza-spoa/etc/coraza.conf: invalid argument"

fionera replied:
Intersting. Will try to reproduce

superstes commented:
Had it running the last 20 days without any major issues: (_with go 1.22.3_)
* Only a few errors per day in relation to the hits/requests
  ![image](https://github.com/user-attachments/assets/c97f3b27-937c-4572-9847-a16697f4cad7)
* Memory and CPU utilization are stable:
  ![image](https://github.com/user-attachments/assets/fb5baff8-d51e-430e-8612-0aab74ac1540)

dknight-bg commented:
I get an error when result processing is enabled ( response_check: true )  : 

**9:49PM PNC Error handling request error="response id is empty"**


```[coraza]
spoe-agent coraza-agent
    messages coraza-req coraza-res
    option var-prefix coraza
    option set-on-error error
    timeout hello      2s
    timeout idle       2m
    timeout processing 500ms
    use-backend coraza-spoa
    log global

spoe-message coraza-req
    args app=str(test_app) src-ip=src src-port=src_port dst-ip=dst dst-port=dst_port method=method path=path query=query version=req.ver headers=req.hdrs body=req.body
    event on-frontend-http-request

spoe-message coraza-res
    args app=str(test_app) id=var(txn.e2e.id) version=res.ver status=status headers=res.hdrs body=res.body
    event on-http-response`
```

dknight-bg commented:
> > I get an error when result processing is enabled ( response_check: true )  :
> > **9:49PM PNC Error handling request error="response id is empty"**
> 
> The ID argument is using the wrong var. It would have to be txn.coraza.id

Thank you, it works correctly with tnx.coraza.id parameter
I got confused by the sample configuration example/haproxy/coraza.cfg from this branch

jptosso commented:
Merged. Work will continue from here; v0.1 was created to keep the old version. This will be tagged as 1.0.0-experimental
